### PR TITLE
Attempt to implement a sane #isSymbolic yet again

### PR DESCRIPTION
**Pull request by shingarov**

### CODE REVIEW

**Comment by janvrany** [comment]
I see where are you going with this, but I think this patch does not make it any saner, just confusing in different way. 

Question is, what `#isSymbolic` mean? 

We - me including - used `isSymbolic` in a sense "does this AST contain variable(s)?" so following:
```
(Bool true | Bool false) isSymbolic
```
should return `false` (and it does, with this patch). However, above is no different from - say - 
```
(1 toInt + 1 toInt) isSymbolic
```
so it should also return `false` - but it does not (not now, not with this patch). To fix this, we need update `Z3AST >> isSymbolic` too, I think.

**Comment by shingarov** [comment]
Well, to my ears, `1 toInt + 1 toInt` is certainly not symbolic.  Something like d6b78a3 then?

**Comment by janvrany** [comment]
I'm not sure about 
[d6b78a3](https://github.com/shingarov/MachineArithmetic/pull/50/commits/d6b78a3251ab0d5e182646e5d8baaa52123be5d2) either. It introduces its own problems:
```
(Bool true | 'x' toBool) isSymbolic " -> true"
```
but 
```
(0 toInt * 'x' toInt) isSymbolic "-> false"
```

This is because the way it is implemented differs. I think we can make it consistent by either using simplify or checking for variables. Something like
```
Bool >> isSymbolic
   | simplified |
   
   simplified := self simplify.
   ^simplified ~~ ctx mkTrue and:[ simplified ~~ ctx mkFalse ] 
```

or 

```
Z3AST >> isSymbolic
        | variables |
 	variables := self variables.
 	^variables notEmpty  

```

Either way, after all these attempts, I'd really like to see a comment in `Z3AST >> isSymbolic` explaining in plain english what `#isSymbolic` means : - )

**Comment by shingarov** [comment]
> (Bool true | 'x' toBool) isSymbolic " -> true"

Fair enough, if we simplify in `Z3AST` then it makes sense to simplify in `Bool`, too.  But that doesn't really solve the problem, because having some *other* examples where `isSymbolic` will not be computed correctly, is unavoidable even in principle, because of things like Rice's Theorem.  The best we can do is answer `false` when we are presented with a decidable case which the algorithm does know now to decide, and `true` to be on the safe side when we are unsure.  For example,

> (0 toInt * 'x' toInt) isSymbolic "-> false"

happens to be one of those "lucky" ones.  The other extreme is `'x' toInt isSymbolic` -- we know for sure it's symbolic.  There will always be some "in the middle" cases where we "don't know".  Implementations of `#isSymbolic` can be more precise or less precise, thus forming an "algebraic ∩-structure" (or more concretely an "information system" in this case).  We demand that the precision monotonically non-decreases with each new version of Z3 or MachineArithmetic.  The sequence does converge to a limit, but the limit is not computable.
